### PR TITLE
Use ServiceProvider for Rand() 

### PR DESCRIPTION
**Pull request by MikeStall**
This PR shows how we can use the new Dependency Injection support from #503 to have Rand() use a rand service, rather than just call Random directly. 

See the new unit test.

### CODE REVIEW

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryMath.cs [hunk 1]
```diff
@@ -842,15 +838,9 @@ public static FormulaValue RandBetween(IRContext irContext, NumberValue[] args)
             lower = Math.Ceiling(lower);
             upper = Math.Floor(upper);
 
-            lock (_randomizerLock)
-            {
-                if (_random == null)
-                {
-                    _random = new Random();
-                }
+            var random = _defaultRandService;
```
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryMath.cs:841]: `runner.GetService<IRandomService>() ?? _defaultRandService`

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryTable.cs [hunk 2]
```diff
@@ -358,12 +358,9 @@ public static FormulaValue Shuffle(IRContext irContext, FormulaValue[] args)
             var table = (TableValue)args[0];
             var records = table.Rows;
 
-            lock (_randomizerLock)
-            {
-                _random ??= new Random();
-            }
+            var random = _defaultRandService;
```
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryTable.cs:361]: Same comment as `RandBetween`: need to attempt to retrieve the service from the runner

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryMath.cs (hunk 1 above)
Comment (MikeStall) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryMath.cs:841]: Right - this just demonstrates for Rand() function. 
Need to apply to these  other functions too.

File: src/libraries/Microsoft.PowerFx.Interpreter/EvalVisitor.cs [hunk 3]
```diff
@@ -37,6 +37,16 @@ public EvalVisitor(CultureInfo cultureInfo, CancellationToken cancel, ReadOnlySy
...
 
+        public T GetService<T>()
+        {
+            if (_runtimeConfig != null)
+            {
+                return (T)_runtimeConfig.GetService(typeof(T));
+            }
+
+            return default;
+        }
```
Comment (LucGenetier) [inline src/libraries/Microsoft.PowerFx.Interpreter/EvalVisitor.cs:48]: I would throw by default

        public T GetService<T>(bool throwOnMissing = true)
        {
            if (_runtimeConfig != null)
            {
                return (T)_runtimeConfig.GetService(typeof(T), throwOnMissing);
            }

            return throwOnMissing ? throw new Exception($"Cannot find {typeof(T).FullName}") : default;
        }

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/IRandomService.cs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        private static readonly Random _random = new Random();
+
+        public virtual double NextDouble()
+        {
+            lock (_randomizerLock)
+            {
+                return _random.NextDouble();
+            }
+        }
+    }
```
Comment (LucGenetier) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/IRandomService.cs:38]: Should be in a different file

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/IRandomService.cs [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    {
+        /// <summary>
+        /// Get the next random number. 
+        /// </summary>
+        /// <returns></returns>
+        double NextDouble();
+    }
+
+    // Default implementation of IRandomService
+    internal class DefaultRandomService : IRandomService
```
Comment (LucGenetier) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/IRandomService.cs:24]: >DefaultRandomService [](http://example.com/codeflow?start=19&length=20)

I would add a comment
Not suitable for any security related purpose.

See [https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca5394](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca5394)

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/RecalcEngineTests.cs [hunk 6]
```diff
@@ -772,6 +773,25 @@ public void TestMultiReturn()
...
+        }
+
+        private class TestRandService : IRandomService
+        {
+            public double NextDouble()
+            {
+                return 555;
+            }
+        }
+
```
Comment (LucGenetier) [inline src/tests/Microsoft.PowerFx.Interpreter.Tests/RecalcEngineTests.cs:794]: That's a random service :D

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/IRandomService.cs (hunk 4 above)
Comment (MikeStall) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/IRandomService.cs:41]: It's helpful to have the interface + default implementation together.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryMath.cs [hunk 7]
```diff
@@ -13,10 +13,13 @@ namespace Microsoft.PowerFx.Functions
     // Direct ports from JScript. 
     internal static partial class Library
     {
-        private static readonly object _randomizerLock = new object();
+        private static readonly IRandomService _defaultRandService = new DefaultRandomService();
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryMath.cs:16]: >IRandomService [](http://example.com/codeflow?start=32&length=14)

This is pushing locking responsibilities to the IRandomService

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryMath.cs [hunk 8]
```diff
@@ -811,20 +814,41 @@ public static FormulaValue Power(IRContext irContext, NumberValue[] args)
...
+            if (value < 0 || value > 1)
+            {
+                // this a bug in the host's IRandomService
+                throw new InvalidOperationException($"IRandomService ({random.GetType().FullName}) returned an illegal value {value}. Must be between 0 and 1");
             }
+
+            return value;
+        }
+
+        private static double SafeNextDouble(this IServiceProvider services)
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryMath.cs:833]: >SafeNextDouble [](http://example.com/codeflow?start=30&length=14)

Now that we are allowing hosts to provide services, we need to have more validation that they're doing it correctly. Else their bugs become our bugs.

**Review by LucGenetier — DISMISSED** [review]

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryMath.cs [hunk 9]
```diff
@@ -811,20 +814,41 @@ public static FormulaValue Power(IRContext irContext, NumberValue[] args)
...
-                if (_random == null)
-                {
-                    _random = new Random();
-                }
+            var value = random.NextDouble(); 
 
-                return new NumberValue(irContext, _random.NextDouble());
+            if (value < 0 || value > 1)
+            {
+                // this a bug in the host's IRandomService
```
Comment (lesaltzm) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryMath.cs:826]: grammar nit: `This is`

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryMath.cs [hunk 10]
```diff
@@ -13,10 +13,13 @@ namespace Microsoft.PowerFx.Functions
...
-        private static readonly object _randomizerLock = new object();
+        private static readonly IRandomService _defaultRandService = new DefaultRandomService();
 
-        [ThreadSafeProtectedByLock(nameof(_randomizerLock))]
-        private static Random _random;
+        private static T GetService<T>(this IServiceProvider services, T defaultService)
+        {
+            var service = (T)services.GetService(typeof(T));
+            return service ?? defaultService;
+        }
```
Comment (lesaltzm) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryMath.cs:22]: This seems like a generally useful helper as we expose more services, is there a better place for it than LibraryMath.cs?

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Review by lesaltzm — APPROVED** [review]
🚢

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Pull request merged**